stdlib/fun.mli

```diff
@@ -31,3 +31,29 @@ val flip : ('a -> 'b -> 'c) -> ('b -> 'a -> 'c)
 val negate : ('a -> bool) -> ('a -> bool)
 (** [negate p] is the negation of the predicate function [p]. For any
     argument [x], [(negate p) x] is [not (p x)]. *)
+
+val protect : acquire:(unit -> 'a) -> release:('a -> unit) -> ('a -> 'b) -> 'b
+(** [protect ~acquire ~release work] invokes [acquires], then [work],
+   and then invokes [release] when [work] returns with its value or an
+   exception. In the latter case the exception is re-raised after
+   [release]. If [release] raises an exception, then the exception
+   [Release_failure] is raised instead (see [Release_failure] below).
+
+    [protect] lets you manage resources reliably, under the following
+   conditions: 1) the acquisition either succeeds, or if it raises an
+   exception then it is without having acquired the resource (strong
+   exception safety); 2) the release of the resource never fails. If
+   the conditions are met, then the resource is guaranteed to have
+   been released when [protect] returns.
+
+    Note: [protect] does not yet protect against asynchronous
+   exceptions raised inside [acquire] or [release] by signal handlers,
+   such as Sys.Break. Furthermore, an asynchronous exception in
+   [release] can be promoted into a [Release_failure] exception.  *)
+
+exception Release_failure of exn
+(** Raised by [protect] when [~release] raises an exception. The
+   argument is the exception that was raised. As a general rule, this
+   exception should not be caught, it denotes a programming error and
+   the code should be modified not to trigger it. *)
+
```

Comment:
The CI [complains](https://travis-ci.org/ocaml/ocaml/jobs/455385167#L477) about this trailing empty line. You should also add a Changes entry for the CI to pass. I guess you should merge it with [that one](https://github.com/ocaml/ocaml/blob/2e5b9d1ef1710c2b7addc39a0905b095f12f070a/Changes#L122-L126).